src/SAWScript/Prover/MRSolver/Solver.hs

```diff
@@ -473,29 +477,59 @@ mrRefines' m1 (MaybeElim (Type (asEq -> Just (tp,e1,e2))) m2 f2 _) =
      if cond_holds then mrRefines m1 m2' else
        withAssumption cond (mrRefines m1 m2') >>
        withAssumption not_cond (mrRefines m1 m2)
-mrRefines' (Ite cond1 m1 m1') m2_all@(Ite cond2 m2 m2') =
-  liftSC1 scNot cond1 >>= \not_cond1 ->
-  (mrEq cond1 cond2 >>= mrProvable) >>= \case
-  True ->
-    -- If we can prove cond1 == cond2, then we just need to prove m1 |= m2 and
-    -- m1' |= m2'; further, we need only add assumptions about cond1, because it
-    -- is provably equal to cond2
-    withAssumption cond1 (mrRefines m1 m2) >>
-    withAssumption not_cond1 (mrRefines m1' m2')
-  False ->
-    -- Otherwise, prove each branch of the LHS refines the whole RHS
-    withAssumption cond1 (mrRefines m1 m2_all) >>
-    withAssumption not_cond1 (mrRefines m1' m2_all)
+
 mrRefines' (Ite cond1 m1 m1') m2 =
-  do not_cond1 <- liftSC1 scNot cond1
-     withAssumption cond1 (mrRefines m1 m2)
-     withAssumption not_cond1 (mrRefines m1' m2)
+  liftSC1 scNot cond1 >>= \not_cond1 ->
+  mrProvable cond1 >>= \cond1_true_pv->
+  mrProvable not_cond1 >>= \cond1_false_pv ->
+  case (cond1_true_pv, cond1_false_pv) of
+    (True, _) -> mrRefines m1 m2
+    (_, True) -> mrRefines m1' m2
+    _ -> withAssumption cond1 (mrRefines m1 m2) >>
+         withAssumption not_cond1 (mrRefines m1' m2)
 mrRefines' m1 (Ite cond2 m2 m2') =
-  do not_cond2 <- liftSC1 scNot cond2
-     withAssumption cond2 (mrRefines m1 m2)
-     withAssumption not_cond2 (mrRefines m1 m2')
--- FIXME: handle sum elimination
--- mrRefines (Either f1 g1 e1) (Either f2 g2 e2) =
+  liftSC1 scNot cond2 >>= \not_cond2 ->
+  mrProvable cond2 >>= \cond2_true_pv->
+  mrProvable not_cond2 >>= \cond2_false_pv ->
+  case (cond2_true_pv, cond2_false_pv) of
+    (True, _) -> mrRefines m1 m2
+    (_, True) -> mrRefines m1 m2'
+    _ -> withAssumption cond2 (mrRefines m1 m2) >>
+         withAssumption not_cond2 (mrRefines m1 m2')
+
+mrRefines' (Either ltp1 rtp1 f1 g1 t1) m2 =
+  liftSC1 scWhnf t1 >>= \t1' ->
+  mrGetDataTypeAssump t1' >>= \mb_assump ->
+  case (mb_assump, asEither t1') of
```

Comment:
Shouldn't the `asEither` cases be first, before the datatype assumptions? I guess it doesn't matter, because we won't ever add a datatype assumption for a `Left` or `Right` term